controllers/mariabackup_controller.go

```diff
@@ -0,0 +1,218 @@
+		mariaDbErr = multierror.Append(mariaDbErr, err)
+
+		return ctrl.Result{}, fmt.Errorf("error getting MariaDB: %v", mariaDbErr)
+	}
+
+	if backup.Spec.MariaDBRef.WaitForIt && !mariaDb.IsReady() {
+		if err := r.patchStatus(ctx, &backup, r.ConditionComplete.PatcherFailed("MariaDB not ready")); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error patching Backup: %v", err)
+		}
+		return ctrl.Result{}, errors.New("MariaDB not ready")
+	}
+
+	var batchErr *multierror.Error
+	err = r.BatchReconciler.Reconcile(ctx, &backup, mariaDb)
+	batchErr = multierror.Append(batchErr, err)
+
+	patcher, err := r.patcher(ctx, err, req.NamespacedName, &backup)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		return ctrl.Result{}, fmt.Errorf("error getting patcher for Backup: %v", err)
+	}
+
+	err = r.patchStatus(ctx, &backup, patcher)
+	batchErr = multierror.Append(batchErr, err)
+
+	if err := batchErr.ErrorOrNil(); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error creating Job: %v", err)
+	}
+
+	podList := &corev1.PodList{}
+
+	// Create a label selector to filter pods with a specific label
+	labelSelector := labels.SelectorFromSet(labels.Set(map[string]string{
+		"backup-ref": backup.Name,
+	}))
+
+	// Create a ListOptions with the label selector
+	listOptions := &client.ListOptions{
+		LabelSelector: labelSelector,
+	}
+
+	max_attempts := 0
+	for len(podList.Items) == 0 && max_attempts < 10 {
+		// List pods in the specified namespace with the label selector
+		err = r.Client.List(ctx, podList, listOptions)
+		if err != nil {
+			// Handle the error
+			return ctrl.Result{}, fmt.Errorf("error getting job: %v", err)
+		}
+		time.Sleep(3 * time.Second)
+		max_attempts++
+	}
+
+	var podIP string = ""
+	// Iterate through the list of pods
+
+	for _, pod := range podList.Items {
+
+		if pod.Status.Phase == "Running" {
+			podIP = pod.Status.PodIP
+			break
+		} else if pod.Status.Phase == "Succeeded" {
+			podIP = ""
+			break
+		}
+	}
+
+	if podIP != "" {
+		// Start mariadb-backup on the primary pod
+		// TODO: make the backup source configurable primary or replicas
+		// TODO: add s3 configuration parameters to get backup to s3 directly
+		cmd := []string{
+			"/bin/sh",
+			"-c",
+			fmt.Sprintf("mariadb-backup --backup --user=root --password=${MARIADB_ROOT_PASSWORD} --compress --stream=mbstream --parallel 1 --target-dir=/tmp | socat -u stdio TCP:%s:4444 &", podIP),
+		}
+
+		execReq := r.RESTClient.
+			Post().
+			Namespace(mariaDb.Namespace).
+			Resource("pods").
+			Name(*mariaDb.Status.CurrentPrimary).
+			SubResource("exec").
+			VersionedParams(&corev1.PodExecOptions{
+				Container: "mariadb",
+				Command:   cmd,
+				Stdin:     true,
+				Stdout:    true,
+				Stderr:    true,
+			}, runtime.NewParameterCodec(r.Scheme))
+
+		exec, err := remotecommand.NewSPDYExecutor(r.RESTConfig, "POST", execReq.URL())
```

Comment:
Love this ! 💯 ❤️ 💯 ❤️ 💯 ❤️ 💯 ❤️ 

Thank you very much much! This gives a lot of flexibility, haven't thought about this before!

Reply:
Not sure about this approach,

From my experience, I can say that exec mechanism depends on a large number of components (primarily apiserver, CRI and kubelet) and if at least one stops working, the execution of the command will be interrupted with an unprediucted result. No logs are collected.

Please consider spawning a new pod or job with the same volume and podAffinity to the target pod.
It should normally work with ReadWriteOnce volumes as well, since the local filesystem can be simple shared with multiple pods across the single node.

Reply:
In order not to be unfounded, here is a working proposal:

```
MASTER=mydb-1

image=$(kubectl get pod -o yaml "$MASTER" -o json | jq -rc '.spec.containers[] | select(.name=="mariadb").image')
volumeMounts=$(kubectl get pod -o yaml "$MASTER" -o json | jq -rc '.spec.containers[] | select(.name=="mariadb").volumeMounts')
env=$(kubectl get pod -o yaml "$MASTER" -o json | jq -rc '.spec.containers[] | select(.name=="mariadb").env')
volumes=$(kubectl get pod -o yaml "$MASTER" -o json | jq -rc '.spec.volumes')
ip=$(kubectl get pod -o yaml "$MASTER" -o json | jq -r '.status.podIP')
cmd="[\"mariadb-backup\", \"--backup\", \"--user=root\", \"--password=\$(MARIADB_ROOT_PASSWORD)\", \"--compress\", \"--target-dir=/tmp\", \"--protocol=tcp\", \"--host=$ip\"]"

overrides="$(
cat <<EOT
{
  "spec": {
    "containers": [
      {
        "image": "$image",
        "name": "mariadb-backup",
        "stdin": true,
        "stdinOnce": true,
        "tty": true,
        "command": $cmd,
        "env": $env,
        "volumeMounts": $volumeMounts
      }
    ],
    "tolerations": [{ "operator": "Exists" }],
    "volumes": $volumes
  }
}
EOT
)"

kubectl run --image "$image" --restart=Never --overrides="$overrides" -ti mariadb-backup
```

Of course you don't need all those `stdin` and `tty` options when running job from operator